IOS/SwiftSeedProject/SwiftSeedProject/CoreDataManager.swift

```diff
@@ -0,0 +1,132 @@
+//
+//  CoreDataManager.swift
+//  SwiftSeedProject
+//
+//  Created by Lucas Pelizza on 3/30/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+class CoreDataManager {
+    
+    typealias CoreDataManagerCompletion = () -> ()
+    private let modelName: String
+    fileprivate let completion: CoreDataManagerCompletion
+    
+    private var persistentStoreURL: NSURL {
+
+        let storeName = "\(modelName).sqlite"
+        let fileManager = FileManager.default
+        
+        let documentsDirectoryURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
+        
+        return documentsDirectoryURL.appendingPathComponent(storeName) as NSURL
+    }
+    
+    private lazy var managedObjectModel: NSManagedObjectModel? = {
+
+        guard let modelURL = Bundle.main.url(forResource: self.modelName, withExtension: "momd") else {
+            return nil
+        }
+        
+        let managedObjectModel = NSManagedObjectModel(contentsOf: modelURL)
+        
+        return managedObjectModel
+    }()
+    
+    private lazy var persistentStoreCoordinator: NSPersistentStoreCoordinator? = {
+        guard let managedObjectModel = self.managedObjectModel else {
+            return nil
+        }
+        
+        let persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel)
+        
+        return persistentStoreCoordinator
+    }()
+    
+    private lazy var privateManagedObjectContext: NSManagedObjectContext = {
+        
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.persistentStoreCoordinator = self.persistentStoreCoordinator
+        
+        return managedObjectContext
+    }()
+    
+    public private(set) lazy var mainManagedObjectContext: NSManagedObjectContext = {
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
+        managedObjectContext.parent = self.privateManagedObjectContext
+        
+        return managedObjectContext
+    }()
+    
+    init(modelName: String, completion: @escaping CoreDataManagerCompletion) {
+        self.modelName = modelName
+        self.completion = completion
+        
+        setupCoreDataStack()
+    }
+    
+    private func setupCoreDataStack() {
+        let _ = mainManagedObjectContext.persistentStoreCoordinator
+        
+        DispatchQueue.global().async {
+            self.addPersistentStore()
+
+            DispatchQueue.main.async { self.completion() }
```

Comment:
I agree with @lpelizzaMS, it is a common situation to update a UI element on completion, although if we use MVVM we should not update UI directly there. I think it should be executed on main thread because it is safer, so in that way, _what_ doing on completion should be responsibility of the developer.